File: leetcode/19-Remove_Nth_Node_From_End_of_List/main.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:  # noqa:501
        # Set a dummy node to point to head
        dummy = ListNode(0, head)
        left = dummy
        # Set right pointer to head
        right = head
        # Move the right pointer Nth steps forward
        for _ in range(n):
            right = right.next

        # Move left and right pointers until the right pointer reaches the
        # last node.
        # At this point, the left pointer will be pointing to the node behind
        # the nth last node.
        while right:
            right = right.next
            left = left.next

        # Relink the left node to the node next to left's next node
        left.next = left.next.next
        return dummy.next

File: leetcode/19-Remove_Nth_Node_From_End_of_List/main.go
package removenthnodefromendoflist

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	p1, p2 := head, head
	// Move P2 n steps forward
	for i := 0; i < n; i++ {
		p2 = p2.Next
	}

	if p2 == nil {
		head = head.Next
		return head
	}

	for {
		if p2.Next == nil {
			tmp := p1.Next
			p1.Next = p1.Next.Next
			tmp.Next = nil
			break
		}
		p2 = p2.Next
		p1 = p1.Next
	}

	return head
}
